packages/aws-cdk/lib/api/logs/find-cloudwatch-logs.ts

```diff
@@ -0,0 +1,108 @@
+import * as cxapi from '@aws-cdk/cx-api';
+import { Mode, SdkProvider, ISDK } from '../aws-auth';
+import { EvaluateCloudFormationTemplate, LazyListStackResources, establishResourcePhysicalName } from '../evaluate-cloudformation-template';
+
+// resource types that have associated CloudWatch Log Groups that should
+// _not_ be monitored
+const IGNORE_LOGS_RESOURCE_TYPES = ['AWS::EC2::FlowLog', 'AWS::CloudTrail::Trail', 'AWS::CodeBuild::Project'];
+
+// resource types that will create a cloudwatch log group
+// with a specific name if one is not provided
+const RESOURCE_TYPES_WITH_IMPLICIT_LOGS: { [key: string]: string } = {
+  'AWS::Lambda::Function': 'FunctionName',
+};
+
+export async function findCloudWatchLogGroups(
+  sdkProvider: SdkProvider,
+  stackArtifact: cxapi.CloudFormationStackArtifact,
+): Promise<{ env: cxapi.Environment, sdk: ISDK, logGroupNames: string[] }> {
+  const resolvedEnv = await sdkProvider.resolveEnvironment(stackArtifact.environment);
+  // TODO: update this to use the lookup role once https://github.com/aws/aws-cdk/pull/18277 is merged
+  const sdk = await sdkProvider.forEnvironment(resolvedEnv, Mode.ForReading);
+
+  const listStackResources = new LazyListStackResources(sdk, stackArtifact.stackName);
+  const evaluateCfnTemplate = new EvaluateCloudFormationTemplate({
+    stackArtifact,
+    parameters: {},
+    account: resolvedEnv.account,
+    region: resolvedEnv.region,
+    partition: (await sdk.currentAccount()).partition,
+    urlSuffix: sdk.getEndpointSuffix,
+    listStackResources,
+  });
+
+  // map of logicalId to CloudFormation type
+  // e.g. 'mylambdaFunction': 'AWS::Lambda::Function'
+  const logicalIdsOfImplicitLogServices = new Map<string, string>();
+  const logGroupNames: string[] = [];
+  const logGroupLogicalIds = new Set<string>();
+  const template = stackArtifact.template as { [key: string]: any };
+
+  // do a first pass at identifying all log groups
+  if ('Resources' in template) {
+    for (const [key, resource] of Object.entries(template.Resources)) {
```

Comment:
I would just do something simpler instead:

```suggestion
  for (const [logicalId, resource] of Object.entries(template.Resources ?? {})) {
```

It will allow you to get rid of the `if`, one level of indentation, and the early return expression in the `else`.